src/trs/rewrite.rs

```diff
@@ -251,10 +251,709 @@ impl TRS {
+    ///
+    /// ```
+    /// # #[macro_use] extern crate polytype;
+    /// # extern crate programinduction;
+    /// # extern crate rand;
+    /// # extern crate term_rewriting;
+    /// # use programinduction::trs::{TRS, Lexicon};
+    /// # use rand::{thread_rng};
+    /// # use term_rewriting::{Context, RuleContext, Signature, parse_rule, parse_term};
+    /// # fn main() {
+    /// let mut sig = Signature::default();
+    ///
+    /// let r = parse_rule(&mut sig, "A(B(x_ x_)) = D B(x_ x_)").expect("parse of A(B(x_ x_)) = D B(x_ x_)");
+    /// let rule = parse_rule(&mut sig, "C(y_) = B(y_ y_)").expect("parse of C(y_) = B(y_ y_)");
+    ///
+    /// let result = TRS::inverse_evaluate_rule_helper(&rule, &r);
+    ///
+    /// assert_eq!(result.unwrap().pretty(&sig), "A(C(y_)) = D C(y_)");
+    /// # }
+    /// ```
+    pub fn inverse_evaluate_rule_helper(rule: &Rule, r: &Rule) -> Option<Rule> {
+        let lhs = TRS::inverse_evaluation_helper(rule, &r.lhs);
+        let mut rhs: Vec<Term> = vec![];
+        for idx in 0..r.rhs.len() {
+            rhs.push(TRS::inverse_evaluation_helper(rule, &r.rhs[idx]));
+        }
+        Rule::new(lhs, rhs)
+    }
+    /// Selects two Rules from the TRS at random and atempts to inverse evaluate one rule on the other, if it
+    /// succeeds it takes that new rule and inserts it imediately after the background.
+    pub fn inverse_evaluate<R: Rng>(&self, rng: &mut R) -> Result<TRS, SampleError> {
+        let mut trs = self.clone();
+        let num_rules = self.len();
+        let num_background = self
+            .lex
+            .0
+            .read()
+            .expect("poisoned lexicon")
+            .background
+            .len();
+        if num_background >= num_rules - 1 {
+            return Ok(trs);
+        }
+        let ref_idx: usize = rng.gen_range(0, num_rules);
+        let mut target_idx: usize = rng.gen_range(num_background, num_rules);
+        while ref_idx == target_idx {
+            target_idx = rng.gen_range(num_background, num_rules);
+        }
+        let mut ref_rule = trs.utrs.rules[ref_idx].clone();
+        let rhs_idx = rng.gen_range(0, ref_rule.rhs.len());
+        let temp_rule = Rule::new(ref_rule.lhs, vec![ref_rule.rhs[rhs_idx].clone()]);
+        if temp_rule == None {
+            return Ok(trs);
+        }
+        ref_rule = temp_rule.unwrap();
+        if ref_rule.lhs.variables().len() != ref_rule.rhs[0].variables().len() {
+            return Ok(trs);
+        }
+        let new_rule = TRS::inverse_evaluate_rule_helper(&ref_rule, &trs.utrs.rules[target_idx]);
+        if new_rule == None {
+            return Ok(trs);
+        }
+        trs.utrs.remove_idx(target_idx).expect("removing old rule");
+        trs.utrs
+            .insert_idx(target_idx, new_rule.unwrap())
+            .expect("inserting new rule");
+        Ok(trs)
+    }
+    // generalizes a rule by one step, converts one constant that exists in both sides of a rule into a variable
```

Comment:
It looks like a lot of the code from here down isn't really being used yet. If these are for features you haven't completed yet, can you please remove until the feature is complete? Thanks!